### PR TITLE
[tools] allow explicitly specifying the JDK to use via a new config setting

### LINKED ISSUE #106416: explicitly set java binary path with flutter config
## Use case
I have to work with multiple java version, and flutter uses gradle 6.7 which is not comfortable with latest java versions,
Alternatively, I have to perform changes on every projects `android\gradle.properties` to work with
```
org.gradle.java.home=C:\\Program Files\\Java\\jdk1.8.0_331\\
```

## Proposal
It will be great if we can set java home or executable with flutter config
```
PS C:\Users\MSI> flutter config --java-binary=<path/to/java/executable>
Setting "java-binary" value to "<path/to/java/executable>".

You may need to restart any open editors for them to read new settings.
```

### PR DESCRIPTION
**Pull request by andrewkolos**
Closes https://github.com/flutter/flutter/issues/106416.

This PR adds a new `flutter config` setting named `jdk-dir`. When set, the tool will use the JDK found at this location for all Java-dependent tool operations such as building Android apps via gradle and running Android SDK tools.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/config.dart [hunk 1]
```diff
@@ -20,6 +22,11 @@ class ConfigCommand extends FlutterCommand {
       help: 'Clear the saved development certificate choice used to sign apps for iOS device deployment.');
     argParser.addOption('android-sdk', help: 'The Android SDK directory.');
     argParser.addOption('android-studio-dir', help: 'The Android Studio install directory. If unset, flutter will search for valid installs at well-known locations.');
+    argParser.addOption('jdk-dir', help: 'The Java Development Kit (JDK) install/home directory. '
+      'If unset, flutter will search for one in the following order:\n'
+      '    1) the JDK bundled with the latest installation of Android Studio,\n'
+      '    2) the JDK found at the directory found in the JAVA_HOME environment variable, and\n'
+      "    3) the directory containing the java binary found in the user's path.");
```
Comment (andrewkolos) [inline packages/flutter_tools/lib/src/commands/config.dart:29]: I'd like to get this ordered list to be indented, but I couldn't find a way to do this. Indenting with spaces or using `\t` doesn't seem to work as this seem to get removed in the output. (The spaces currently here are there to satisfy args_test.dart). Here is the output from `flutter config`:

```
    --android-sdk                               The Android SDK directory.
    --android-studio-dir                        The Android Studio install directory. If unset, flutter will search for valid installs at well-known locations.
    --jdk-dir                                   The Java Development Kit (JDK) install/home directory. If unset, flutter will search for one in the following order:
                                                1) the JDK bundled with the latest installation of Android Studio,
                                                2) the JDK found at the directory found in the JAVA_HOME environment variable, and
                                                3) the directory containing the java binary found in the user's path.
    --build-dir=<out/>                          The relative path to override a projects build directory.
    --[no-]enable-web                           Enable or disable Flutter for web.
```

File: packages/flutter_tools/lib/src/commands/config.dart [hunk 2]
```diff
@@ -20,6 +22,11 @@ class ConfigCommand extends FlutterCommand {
       help: 'Clear the saved development certificate choice used to sign apps for iOS device deployment.');
     argParser.addOption('android-sdk', help: 'The Android SDK directory.');
     argParser.addOption('android-studio-dir', help: 'The Android Studio install directory. If unset, flutter will search for valid installs at well-known locations.');
+    argParser.addOption('jdk-dir', help: 'The Java Development Kit (JDK) install/home directory. '
```
Comment (andrewkolos) [inline packages/flutter_tools/lib/src/commands/config.dart:25]: The `-dir` on the end of the name seems a bit redundant/unnecessary, but it matches convention the prexisting `android-studio-dir` and `build-dir` option (which `android-sdk` annoyingly does not follow).

I suppose we could always unify these and then supply the old names in the `aliases` argument to maintain backwards compatibility. I digress.

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/commands/config.dart:25]: unifying while not breaking by using aliases SGTM

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/commands/config.dart:24]: I think you can remove the trailing space.

Also, `"install/home directory"` seems needlessly ambiguous. Can we just say `"install directory."`?

File: packages/flutter_tools/lib/src/commands/config.dart [hunk 3]
```diff
@@ -154,6 +160,10 @@ class ConfigCommand extends FlutterCommand {
       _updateConfig('android-studio-dir', stringArg('android-studio-dir')!);
     }
 
+    if (argResults?.wasParsed('jdk-dir') ?? false) {
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/config.dart:163]: I think here (and throughout this function) we should just cast `argResults` to a non-nullable value and get rid of all the nullable fallbacks.

File: packages/flutter_tools/lib/src/commands/config.dart (hunk 2 above)
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/commands/config.dart:24]: > I think you can remove the trailing space.

The trailing space is necessary to satisfy [missing_whitespace_between_adjacent_strings](https://dart-lang.github.io/linter/lints/missing_whitespace_between_adjacent_strings.html)

> Also, `"install/home directory"` seems needlessly ambiguous.

Agreed. Updated to 'installation directory'.

File: packages/flutter_tools/lib/src/commands/config.dart [hunk 4]
```diff
@@ -203,6 +213,10 @@ class ConfigCommand extends FlutterCommand {
     if (results['android-sdk'] == null && androidSdk != null) {
       results['android-sdk'] = androidSdk.directory.path;
     }
+    final Java? java = globals.java;
+    if (results['jdk-dir'] == null && java != null) {
+      results['jdk-dir'] = java.javaHome;
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/config.dart:218]: this is actually surprising to me, but I suppose we shouldn't break this.

File: packages/flutter_tools/lib/src/android/android_workflow.dart [hunk 5]
```diff
@@ -288,6 +292,7 @@ class AndroidLicenseValidator extends DoctorValidator {
 
   Future<bool> _checkJavaVersionNoOutput() async {
     final String? javaBinary = Java.find(
+      config: _config,
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/android/android_workflow.dart:295]: do we need to call this function again if we already have a `final Java? _java;` in scope?

File: packages/flutter_tools/test/commands.shard/hermetic/config_test.dart [hunk 6]
```diff
@@ -70,6 +70,9 @@ void main() {
 
         expect(jsonObject.containsKey('android-sdk'), true);
         expect(jsonObject['android-sdk'], isNotNull);
+
+        expect(jsonObject.containsKey('jdk-dir'), true);
+        expect(jsonObject['jdk-dir'], isNotNull);
```
Comment (christopherfujino) [inline packages/flutter_tools/test/commands.shard/hermetic/config_test.dart:75]: for these and the other `isNotNull` expectations in this test, can we expect on the actual value, to verify we're not letting the real implementations leak through?

File: packages/flutter_tools/lib/src/android/android_workflow.dart (hunk 5 above)
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/android/android_workflow.dart]: Good catch. Removed with commit [remove extraneous call to Java.find](https://github.com/flutter/flutter/pull/128264/commits/8300ebbd7c6a5b8880d32a4ec0f5dacf66b2578d)

File: packages/flutter_tools/test/general.shard/android/java_test.dart [hunk 7]
```diff
@@ -127,6 +134,37 @@ OpenJDK 64-Bit Server VM Zulu19.32+15-CA (build 19.0.2+7, mixed mode, sharing)
...
+          ),
+        );
+
+        final FakePlatform platformWithJavaHome = FakePlatform(
+          environment: <String, String>{
+            'JAVA_HOME': '/old/jdk'
+          },
+        );
+
+        final Java? java = Java.find(
```
Comment (christopherfujino) [inline packages/flutter_tools/test/general.shard/android/java_test.dart:155]: Can you actually call `Java.find()` twice in this test, once before `config.setValue('jdk-dir', ...)` and once after, expecting the results are different?

File: packages/flutter_tools/lib/src/commands/config.dart (hunk 2 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/commands/config.dart]: > The trailing space is necessary to satisfy [missing_whitespace_between_adjacent_strings](https://dart-lang.github.io/linter/lints/missing_whitespace_between_adjacent_strings.html)

Whoops, good thing it was there too, totally misread this.

File: packages/flutter_tools/test/general.shard/android/java_test.dart (hunk 7 above)
Comment (andrewkolos) [reply packages/flutter_tools/test/general.shard/android/java_test.dart:155]: Isn't this already implicitly covered by the other tests?

File: packages/flutter_tools/test/commands.shard/hermetic/config_test.dart (hunk 6 above)
Comment (andrewkolos) [reply packages/flutter_tools/test/commands.shard/hermetic/config_test.dart]: Done with commit [make machine flag test more precise](https://github.com/flutter/flutter/pull/128264/commits/f45814e3b5d288cbb00dd93c69908e50135caf2a).

This made me realize the inconsistency with some of these fields. `AndroidStudio.directory` is a string, `AndroidSdk.directory` is a `Directory`, and `Java.javaHome` doesn't fit into the `<class>.directory` scheme.

File: packages/flutter_tools/lib/src/android/java.dart [hunk 8]
```diff
@@ -181,10 +184,16 @@ class Java {
...
 
 String? _findJavaHome({
+  required Config config,
   required Logger logger,
   required AndroidStudio? androidStudio,
   required Platform platform,
 }) {
+  final Object? configured = config.getValue('jdk-dir');
+  if (configured != null) {
+    return configured as String;
```
[the remark this answers is not in the document]
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/android/java.dart:194]: Filed [[tools] Consider adding getter methods to Config that cast values for convenience #128278
](https://github.com/flutter/flutter/issues/128278)

File: packages/flutter_tools/test/general.shard/android/java_test.dart (hunk 7 above)
Comment (christopherfujino) [reply packages/flutter_tools/test/general.shard/android/java_test.dart:155]: We can only rely on other tests verifying the exact setup of this test if we know that now and forever this test and the dependent test are set up in exactly the same way. Also that that other test is never deleted.

In other words, I would say that this test is not completely verifying that `Java.find()` is finding `java.javaHome` via the `Config`.

**Review by christopherfujino — APPROVED** [review]
LGTM

**Pull request merged**